File: Python/37. Sudoku Solver.py
'''
回溯：横向遍历列和行
'''


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """

        def isValid(board, row, col, num):
            for i in range(9):
                if board[i][col] == num:
                    return False
            for j in range(9):
                if board[row][j] == num:
                    return False
            block_row = row // 3
            block_col = col // 3
            for i in range(3 * block_row, 3 * block_row + 3):
                for j in range(3 * block_col, 3 * block_col + 3):
                    if board[i][j] == num:
                        return False
            return True

        def backtrack(board):
            for i in range(9):
                for j in range(9):
                    if board[i][j] == '.':
                        for num in range(1, 10):
                            if isValid(board, i, j, str(num)):
                                board[i][j] = str(num)
                                if backtrack(board): return True
                                board[i][j] = '.'
                        return False
            return True

        backtrack(board)

File: C++/37. Sudoku Solver.cpp
/*
回溯：横向遍历列和行
*/
class Solution {
public:
    bool isValid(const vector<vector<char>>& board, int row, int col, char num) {
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == num) return false;
        }
        for (int j = 0; j < 9; j++) {
            if (board[row][j] == num) return false;
        }
        int block_row = row / 3;
        int block_col = col / 3;
        for (int i = block_row * 3; i < block_row * 3 + 3; i++) {
            for (int j = block_col * 3; j < block_col * 3 + 3; j++) {
                if (board[i][j] == num) return false;
            }
        }
        return true;
    }

    bool backtrack(vector<vector<char>>& board) {
        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                if (board[row][col] == '.') {
                    for (char num = '1'; num <= '9'; num++) {
                        if (isValid(board, row, col, num)) {
                            board[row][col] = num;
                            if (backtrack(board)) return true;
                            board[row][col] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    void solveSudoku(vector<vector<char>>& board) {
        backtrack(board);
    }
};